src/iceberg/util/decimal.cc

```diff
@@ -0,0 +1,1044 @@
+  return Divide(multiplier);
+}
+
+Result<Decimal> Decimal::FromBigEndian(const uint8_t* bytes, int32_t length) {
+  static constexpr int32_t kMinDecimalBytes = 1;
+  static constexpr int32_t kMaxDecimalBytes = 16;
+
+  int64_t high, low;
+
+  if (length < kMinDecimalBytes || length > kMaxDecimalBytes) {
+    return InvalidArgument(
+        "Decimal::FromBigEndian: length must be in the range [{}, {}], was {}",
+        kMinDecimalBytes, kMaxDecimalBytes, length);
+  }
+
+  // Bytes are coming in big-endian, so the first byte is the MSB and therefore holds the
+  // sign bit.
+  const bool is_negative = static_cast<int8_t>(bytes[0]) < 0;
+
+  // 1. Extract the high bytes
+  // Stop byte of the high bytes
+  const int32_t high_bits_offset = std::max(0, length - 8);
+  const auto high_bits = UInt64FromBigEndian(bytes, high_bits_offset);
+
+  if (high_bits_offset == 8) {
+    // Avoid undefined shift by 64 below
+    high = high_bits;
+  } else {
+    high = -1 * (is_negative && length < kMaxDecimalBytes);
+    // Shift left enough bits to make room for the incoming int64_t
+    high = SafeLeftShift(high, high_bits_offset * CHAR_BIT);
+    // Preserve the upper bits by inplace OR-ing the int64_t
+    high |= high_bits;
+  }
+
+  // 2. Extract the low bytes
+  // Stop byte of the low bytes
+  const int32_t low_bits_offset = std::min(length, 8);
+  const auto low_bits =
+      UInt64FromBigEndian(bytes + high_bits_offset, length - high_bits_offset);
+
+  if (low_bits_offset == 8) {
+    // Avoid undefined shift by 64 below
+    low = low_bits;
+  } else {
+    // Sign extend the low bits if necessary
+    low = -1 * (is_negative && length < 8);
+    // Shift left enough bits to make room for the incoming int64_t
+    low = SafeLeftShift(low, low_bits_offset * CHAR_BIT);
+    // Preserve the upper bits by inplace OR-ing the int64_t
+    low |= low_bits;
+  }
+
+  return Decimal(high, static_cast<uint64_t>(low));
+}
+
+Result<Decimal> Decimal::Rescale(int32_t orig_scale, int32_t new_scale) const {
+  if (orig_scale < 0 || orig_scale > kMaxScale) {
+    return InvalidArgument(
+        "Decimal::Rescale: original scale must be in the range [0, {}], "
+        "was {}",
+        kMaxScale, orig_scale);
+  }
+  if (new_scale < 0 || new_scale > kMaxScale) {
+    return InvalidArgument(
+        "Decimal::Rescale: new scale must be in the range [0, {}], "
+        "was {}",
+        kMaxScale, new_scale);
+  }
+
+  if (orig_scale == new_scale) {
+    return *this;
+  }
+
+  const int32_t delta_scale = new_scale - orig_scale;
+  const int32_t abs_delta_scale = std::abs(delta_scale);
+  Decimal out;
+
+  assert(abs_delta_scale <= kMaxScale);
+
+  auto& multiplier = kDecimal128PowersOfTen[abs_delta_scale];
+
+  const bool rescale_would_cause_data_loss =
+      RescaleWouldCauseDataLoss(*this, delta_scale, multiplier, &out);
+
+  if (rescale_would_cause_data_loss) {
+    return Invalid("Rescale would cause data loss: {} -> {}", ToIntegerString(),
```

Comment:
Isn't `out.ToIntegerString()` undefined in the error message?

Reply:
Good catch, change to a safe error message.